### PR TITLE
Improve DOSBox Staging application icons

### PR DESCRIPTION
**Pull request by johnnovak**
# Description

Continuation of https://github.com/dosbox-staging/dosbox-staging/pull/2232. Thanks to @ThomasEricB for the idea 🎉 

This is about removing the not-so-good-looking circle from the icons and creating separate icons per OS that adhere to the platform's UI/UX guidelines.

I also took the opportunity to simplify the icon Makefile. Previously, it was a bit overengineered for no good reason, and full of "clever" stuff. I want build scripts to be simple and stupid, so I made it so, also removing extra fluff in the process and stripping it down to bare essentials.

The approach to re-generating the icons has changed too: the derived smaller PNG files and all other derived icon files are now also checked into the repo. The make targets should only be used going forward when we need to re-generate these checked-in assets. The exception is the `install` target for Linux, which I left alone.

These small binary files almost never change, and they're minuscule... The whole "don't check binaries into Git" thing doesn't apply here, of course... That's for big and/or frequently updated files.

# Manual testing

### macOS

Tested the macOS DMG file built by CI—all looks fine (see https://github.com/dosbox-staging/dosbox-staging/pull/3157#issuecomment-1831311959).

### Windows

Tested on Win 10 (see https://github.com/dosbox-staging/dosbox-staging/pull/3157#issuecomment-1835633753 and https://github.com/dosbox-staging/dosbox-staging/pull/3157#issuecomment-1837132798)

### Linux

Partially done.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [ ] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [x] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by johnnovak** [comment]
@Grounded0 Hold your horses, it's a work in progress :) But it will be finished in a day or two.

**Comment by ThomasEricB** [comment]
I wish you kept some of my work (since it was satisfactory before), but oh well

**Comment by johnnovak** [comment]
@Burrito78 @kklobe @kcgen @Grounded0 

The macOS icons are now done and are pixel-perfect. Was fooling around with some alignment issues because my first attempt was slightly off center by a few pixels and it triggered my OCD badly 😅 

You can download "dosbox-staging-macOS-universal " from the bottom of this page to test:

https://github.com/dosbox-staging/dosbox-staging/actions/runs/7028615905?pr=3157

Now on to the rest!

## Installer

image

## Small-size legibility

This is how it looks in my Applications folder:

image

## App switcher

image

**Comment by Grounded0** [comment]
Looking good. Windows and Linux icons next.

**Comment by Burrito78** [comment]
The icon works fine on macOS with the light color scheme but there is an issue when using the dark color scheme.

Maybe adding an outline like on the old icon could be a fix there?
Bildschirmfoto 2023-11-29 um 13 00 42

**Comment by Grounded0** [comment]
We could avoid colour bleeding issues against dark backgrounds by actually having a deep black outline like in the DOS prompt.

**Comment by johnnovak** [comment]
IMO, a light outline would look pretty shit. None of these dark icons on your example have such a thing and they're fine. Also, they're all over the place, meaning we can do anything we like, there's no "uniformity" anyway.

image

I like best how the Mission Control app looks, so make it basically "true 100% black" aka `#000000`, as @Grounded0 suggested. Currently it's only about a 70% black, not 100% black

**Comment by Grounded0** [comment]
Yeah we'll go with Mission Control.app derivative.

**Comment by Grounded0** [comment]
Do you have the deep black icon to show?

**Comment by johnnovak** [comment]
> Can we see that deep black outline soonish so work can progress to Windows/Linux icon. Screenshots against macOS dark mode grey so its easier to wave through.

Later today man.

**Comment by johnnovak** [comment]
Behold! Black as night! 😎 

Get it from here, fresh from the oven 🤘🏻 
https://github.com/dosbox-staging/dosbox-staging/actions/runs/7043758252?pr=3157

image

image

image

image


---

Still looks nice with the light theme, of course:

image

**Comment by Grounded0** [comment]
Looks good to me on all backgrounds. On to Windows/Linux icon.

Pinging to @Burrito78 for final approval.

Lets not pontificate this too much since its much better than what is replaced.

**Review by Burrito78 — APPROVED** [review]
macOS icon looks good, in can't check Windows/Linux so I'm reviewing this now.

**Comment by johnnovak** [comment]
> macOS icon looks good, in can't check Windows/Linux so I'm reviewing this now.

Not ready yet, I'll tell you when it's ready for testing.

**Comment by johnnovak** [comment]
> I'll test it on Wine so its at least better than no testing.
> 
> Do we have a Windows user here to ping?

I use Windows, so I'll test it on Win 10, of course.

What we need is some Linux user (when it's done).

**Comment by Grounded0** [comment]
I got one Linux tester from the depths of IRC once this turns green.

File: contrib/icons/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+- 16x16 - Used in the window's title bar
+- 24x24 - Taskbar icon; if missing, Windows will pick the wrong icon for
+          the window's titlebar as well (it will scale down the biggest icon
+          it can find, so the result will be blurry).
+
+The icon file stored in the repo is packages into the final executable via
+Visual Studio, so make sureo t update `src/winres.rc` is you change the name
+of path of the `.ico` file.
+
```
Comment (kcgen) [inline contrib/icons/README.md:100]: Awesome docs, @johnnovak :+1:

**Review by kcgen — APPROVED** [review]
Looking great @johnnovak, and seconding John's thanks to @ThomasEricB for kicking this off and sharing lots of very nice options almost a year ago.

**Comment by johnnovak** [comment]
> Looking great @johnnovak, and seconding John's thanks to @ThomasEricB for kicking this off and sharing lots of very nice options almost a year ago.

Thanks, but still a draft and Linux / Win stuff is coming today. 😄

**Comment by Grounded0** [comment]
I'll do my monthly funding and this as a bulk payment once this is finished so i don'y have to go twice through the Open Collective payments UI. You can collect payment for this off the OC then. I'll also do funding through 2024 but at a lower level since i have greater personal expenses though 2024.

**Comment by johnnovak** [comment]
> I'll do my monthly funding and this as a bulk payment once this is finished so i don'y have to go twice through the Open Collective payments UI. You can collect payment for this off the OC then. I'll also do funding through 2024 but at a lower level since i have greater personal expenses though 2024.

Sounds good man. Thanks! 😄

File: contrib/icons/README.md (hunk 1 above)
Comment (Torinde) [reply contrib/icons/README.md:100]: sureo t -> sure to
is you change -> if you change

**Comment by johnnovak** [comment]
Yeah, also fooling around to simplify this whole thing as it's a bit over-engineered & make the Linux build work again...

**Comment by Grounded0** [comment]
https://github.com/dosbox-staging/dosbox-staging/blob/main/contrib/icons/dosbox-staging.bmp

**Comment by Grounded0** [comment]
`make -C contrib/icons/ linux`

Should be:

`make -C contrib/icons/linux`

**Comment by johnnovak** [comment]
Ok, in theory I'm done with everything. Linux is done too but needs testing.

I'll clean up the commit log tomorrow then we're done once the Linux results have been confirmed.

**Comment by Burrito78** [comment]
> Dispatching Linux tester.
> 
> 
> 
> About file `contrib/macos/DS_Store`.
> 
> 
> 
> Knowing something about macOS i think that file shouldn't be there.

That's a file from Synology Disk Station Manager, a NAS OS.

**Comment by Grounded0** [comment]
Doesn't seem to be smooth sailing with KDE Plasma.

Installed versions of the desktop environment stack:

X.org 21.1.9. Qt 5.15.11. KDE Plasma 5.27.9.

The bitmaps seem to get picked just fine but the big SVG icon isn't picked up at all. The icon cache has been attempted to be flushed by exiting the X11 session.

**Comment by Grounded0** [comment]
Pinging @kcgen as this falls into the persons domain.

**Comment by Grounded0** [comment]
The main problem blocking deployment:

KDE Plasma:

Incorrect icons in the title bar and task switcher. Those are DEFINITELY Staging's problem, as there Plasma just picks up whatever the program sets.

Relevant function:

https://wiki.libsdl.org/SDL2/SDL_SetWindowIcon

Any hand tuning of the small pixel grids can be done in a follow-up PRs after the main aim of eliminating ugly has been achieved with this.

**Comment by johnnovak** [comment]
Linux screenshots of the problem, please. It's also unclear if this Linux issue is a regression compared to 0.80.1 or something new.

Our "official" Linux downloads of 0.80.1 must be used for regression testing the icons. My understanding is that Staging installed from any random distros package manager is different; the distro packagers can do any number of changes there, including changing the way the icons behave.

I just want to achieve parity with 0.80.1 in this PR; further Linux related icon enhancements will be the task for someone who uses Linux as their primary OS.

@weirddan455 @kcgen @FeralChild64 can any of you help test the icons using the Linux artifact from this PR, _and_ regression test the behaviour using our official 0.80.1 Linux packages? Thanks!

@PatTheMav 

The border need to be there even on the small icons as people could use a theme with 100% white colours and then the icons would blend into that without outline. So it stays.

Regarding small icon sizes, the 32x32 and smaller icons are generated from tweaked SVG that result in somewhst better pixel grid alignment. This could be probably improved, but yeah, later, and TBH I don't particularly care 😄 I _might_ break out Deluxe Paint _one day_ to draw small icons by hand, but don't count on it 😎 If anyone wants to do this in the future, they're most welcome.

**Comment by weirddan455** [comment]
>  The bitmaps seem to get picked just fine but the big SVG icon isn't picked up at all. The icon cache has been attempted to be flushed by exiting the X11 session.

I use KDE Plasma.  Repro instructions and screenshot please.  I don't know how you're able to tell which icon is being loaded and where exactly you're seeing the wrong icon.

Also, try manually removing `~/.cache/icon-cache.kcache` if you want to clear the icon cache.

**Comment by johnnovak** [comment]
I fucked up the SVG copy command for Linux in the Makefile, mea culpa 😊 

Fix & final PR coming soon.

**Comment by johnnovak** [comment]
All green and I'm pretty confident I fixed the SVG icon issue on Linux. Could you please re-test @weirddan455 @Grounded0 using the latest Linux artifacts?

To "trigger" the SVG icon, you could set the icon size in the file explorer to "huge", as big as you can go with it, then make sure the icon is not pixelated.

Can you also please paste screenshot evidence of your findings? Thanks!

**Comment by Grounded0** [comment]
@dreamer added SVG as an inline resource to file `src/gui/icon.c`. This probably should be fixed since we don't want icon data hardcoded into our `.c` files exactly for this reason. This is being called up by the SDL as evidenced in these screenshots:

KDE Plasma uses quite limited Qt SVG module for rasterisation, which is likely why the line is so thick.

This is how the icons look in the system, fresh after make install command:

image

This is how it looks if the user manually deletes all PNGs and leave only the SVG:

image

The window, irrespective of whether I have icons installed at all:

image

Large icons task switcher, same thing:

image

And finally, the taskbar, with all the icons installed:

image

**Comment by johnnovak** [comment]
@Grounded0 

> KDE Plasma uses quite limited Qt SVG module for rasterisation, which is likely why the line is so thick.

No, I know what that is. I'm using some feature not supported by all SVG rasterisers, apparently. I know what the workaround it, stay tuned.

> dreamer added SVG as an inline resource to file src/gui/icon.c. This probably should be fixed since we don't want icon data hardcoded into our .c files exactly for this reason. 

Fuck me, that's some weird shit 😐 This PR keeps giving... I will try to get rid of such nastiness and only reference external files, yeah.

**Comment by weirddan455** [comment]
We need to be using external icon files really.  `SDL_SetWindowIcon()` is a no-op on Linux + Wayland.  The correct way to set the icon on Wayland is to ship a .desktop file that points to the icon we want to use.

**Comment by johnnovak** [comment]
> We need to be using external icon files really. `SDL_SetWindowIcon()` is a no-op on Linux + Wayland. The correct way to set the icon on Wayland is to ship a .desktop file that points to the icon we want to use.

I think this needs someone who has experience on how icons are handled on Linux which I lack, plus I'm not passionate about Linux either...

I suggest I fix the SVG thick line problem (fix coming soon for that), then we merge this. Further Linux icon fixes would be additional enhancements on top of this work.

Reality check: 95%+ of our users are on Windows, and the Windows icons are fine. The macOS icons are also in a very good shape. The Linux situation is about halfway done, but maybe we 1-2% of our users are on Linux, at best.

**Comment by Grounded0** [comment]
@johnnovak 

Agreed. Usually people find motivation to help us only after we blow up something from them anyway.

**Comment by weirddan455** [comment]
>  I think this needs someone who has experience on how icons are handled on Linux which I lack, plus I'm not passionate about Linux either...

Yeah, I can probably tackle the Linux side.  I mainly use X11 on Linux which that code works fine for.  Wayland doesn't work so well with my old Nvidia card.  I can test in Wayland though using open source drivers.

A .desktop file is a shortcut basically.  It's just a text file.  We have one in `contrib/linux/dosbox-staging.desktop`.  Just needs testing.

Of course, if you get rid of the emedded icon and use all external files, that would make things go smoother.

**Comment by johnnovak** [comment]
## Windows installer

image

image

**Comment by johnnovak** [comment]
Once my last commit turns green, I'm merging this good thing. Then refining the Linux situation is a job for later if @weirddan455 or someone is up for it.

**Comment by Grounded0** [comment]
Can you see new icon in `main` .dmg package with macOS?

**Pull request merged**